src/aleph/http/http2.clj

```diff
@@ -0,0 +1,503 @@
+          (instance? byte-array-class body))
+        (send-contiguous-body ch stream headers body)
+
+        (instance? ChunkedInput body)
+        (send-chunked-body ch stream headers body)
+
+        (instance? RandomAccessFile body)
+        (send-chunked-body ch stream headers
+                           (ChunkedFile. ^RandomAccessFile body
+                                         ^int file-chunk-size))
+
+        (instance? File body)
+        (send-chunked-body ch stream headers
+                           (ChunkedNioFile. ^File body ^int file-chunk-size))
+
+        (instance? Path body)
+        (send-chunked-body ch stream headers
+                           (ChunkedNioFile. ^FileChannel
+                                            (FileChannel/open
+                                              ^Path body
+                                              (into-array OpenOption [StandardOpenOption/READ]))
+                                            ^int file-chunk-size))
+
+        (instance? HttpFile body)
+        (send-http-file ch stream headers body)
+
+        (instance? FileChannel body)
+        (send-chunked-body ch stream headers
+                           (ChunkedNioFile. ^FileChannel body ^int file-chunk-size))
+
+        (instance? FileRegion body)
+        (if (or (-> ch (.pipeline) (.get ^Class SslHandler))
+                (some-> ch (.parent) (.pipeline) (.get ^Class SslHandler)))
+          (let [emsg (str "FileRegion not supported with SSL in Netty")
+                ex (ex-info emsg {:ch ch :headers headers :body body})
+                e (Http2FrameStreamException. stream Http2Error/PROTOCOL_ERROR ex)]
+            (log/error e emsg)
+            (netty/close ch))
+          (send-file-region ch headers body))
+
+        :else
+        (try
+          (send-streaming-body ch stream headers body chunk-size)
+          (catch Throwable e
+            (log/error e "Error sending body of type " (StringUtil/simpleClassName body))
+            (throw e))))
+
+      (catch Exception e
+        (log/error e "Error sending message")
+        (throw (Http2FrameStreamException. (.stream ch)
+                                           Http2Error/PROTOCOL_ERROR
+                                           (ex-info "Error sending message" {:headers headers :body body} e)))))))
+
+;; NOTE: can't be as vague about whether we're working with a channel or context in HTTP/2 code,
+;; because we need access to the .stream method. We have a lot of code in aleph.netty that
+;; branches based on the class (channel vs context), but that's not ideal. It's slower, and
+;; writing to the channel vs the context means different things, anyway, they're not
+;; usually interchangeable.
+(defn send-request
+  [^Http2StreamChannel ch req response]
+  (log/trace "http2 send-request fired")
+
+  (when (multipart/is-multipart? req)
+    ;; shouldn't reach at the moment, but just in case
+    (throw (ex-info "HTTP/2 code path multipart not supported"
```

Comment:
Maybe, but it'll be caught right above in `http2-req-handler`. Honestly, I was planning to delete this code, since it's literally checked right before being called.

Reply:
Removed.